File: binding/lua_c/testSearcher.lua
--[[
ip2region lua client test script

@author chenxin<chenxin619315@gmail.com>
]]--

-- check the command line arguments
if ( not arg[1] ) then
    print([[
Usage: lua testSearcher.lua [ip2region db file] [algorithm]
+-Optional Algorithm: binary, b-tree, memory]]);
    os.exit();
end

local Ip2region = require "Ip2region";
-- local cjson = require "cjson";
-- local socket = require "socket";


-- check and parse the dbFile and the method algorithm
-- Create a new ip2region object by the new interface
local ip2region = Ip2region.new(arg[1]);

-- reset the dbFile by the follow two ways:
-- ip2region.dbFile = arg[1];
-- ip2region:setDbFile(arg[1]);


local algorithm = "btree";
if ( arg[2] ~= nil ) then
    local arg_2 = string.lower(arg[2]);
    if ( arg_2 ~= "binary" and arg_2 ~= "memory" ) then
        algorithm = "binary";
    elseif ( arg_2 == "memory" ) then
        algorithm = "memory";
    end
end



-- local data = searcher:memorySearch("120.79.17.142");
-- local data = searcher:binarySearch("120.79.17.142");
-- local data = searcher:btreeSearch("120.79.17.142");
print("initializing " .. algorithm ..[[ 
+----------------------------------+
| ip2region test script            |
| Author: chenxin619315@gmail.com  |
| Type 'quit' to exit program      |
+----------------------------------+]]
);

while ( true ) do
    io.write("ip2region>> ");
    io.input(io.stdin);
    local line = io.read();
    if ( line == nil ) then
        -- do nothing
        break;
    elseif ( line == "quit" ) then
        break;
    elseif ( ip2region.ip2long(line) == nil ) then
        print("Invalid ip address=", line);
    else
        local data;
        local s_time = os.clock();
        if ( algorithm == "btree" ) then
            data = ip2region:btreeSearch(line);
        elseif ( algorithm == "binary" ) then
            data = ip2region:binarySearch(line);
        elseif ( algorithm == "memory" ) then
            data = ip2region:memorySearch(line);
        end

        local cost_time = (os.clock() - s_time) * 1000; -- to millseconds
        if ( data == nil ) then
            io.write("Failed for ip=", line, " is it a valid ip address ?");
        else
            io.write(string.format("%u|%s in %5f millseconds\n", data.city_id, data.region, cost_time));
        end
    end
end

-- close the object
-- print(ip2region);
ip2region:close();

File: v1.0/binding/python/testSearcher.py
#-*- coding:utf-8 -*-
"""
" ip2region python seacher client module
"
" Autho: koma<komazhang@foxmail.com>
" Date : 2015-11-06
"""
import struct, sys, os, time
from platform import python_version

from ip2Region import Ip2Region

def testSearch():
    """
    " ip2region test function
    """
    argLen     = len(sys.argv)
    version    = python_version()
    algorithms = ["binary", "b-tree", "memory"]

    if argLen < 2:
        print("Usage: python testSearcher.py [ip2region db file] [alrogrithm]")
        print("Algorithm: %s" % ", ".join(algorithms))
        return 0

    dbFile = sys.argv[1]

    if (not os.path.isfile(dbFile)) or (not os.path.exists(dbFile)):
        print("[Error]: Specified db file is not exists.")
        return 0

    if argLen > 2:
        algorithm = sys.argv[2]
    try:
        algorithms.index(algorithm)
    except Exception as e:
        algorithm = "b-tree"

    print("initializing %s..." % (algorithm))
    print("+----------------------------------+")
    print("| ip2region test program           |")
    print("| Author: chenxin619315@gmail.com. |")
    print("| Type 'quit' to exit program      |")
    print("+----------------------------------+")

    searcher = Ip2Region(dbFile)

    while True:
        if version[:1] == "2":
            line = raw_input("ip2region>> ")
        else:
            line = input("ip2region>> ")
        line = line.strip()

        if line == "":
            print("[Error]: Invalid ip address.")
            continue

        if line == "quit":
            print("[Info]: Thanks for your use, Bye.")
            break

        if not searcher.isip(line):
            print("[Error]: Invalid ip address.")
            continue

        try:
            sTime = time.time()*1000
            if algorithm == "binary":
                data = searcher.binarySearch(line)
            elif algorithm == "memory":
                data = searcher.memorySearch(line)
            else:
                data = searcher.btreeSearch(line)
            eTime = time.time()*1000
            print("%s|%s in %5f millseconds" % (data["city_id"], data["region"].decode('utf-8'), eTime - sTime))
        except Exception as e:
            print("[Error]: %s" % e)

    searcher.close()

if __name__ == "__main__":
    testSearch()
